### PR TITLE
[WIP] - Client side load balancing

### PR DESCRIPTION
**Pull request by sebaraba**
### Done:
- [x] Multiple http node support;
- [x] Old tests are running;
- [x] New test for multiple nodes;

### Mentions with respect to the comments: 
- [x] I had to pull in `release/4.0` branch in order to overcome the `CLEARTEXT error` and be able to run the tests.

- [x]  I noticed there is a rule which enforced me to keep the url/urls parameter on the last position in the constructor:
```java
public HttpService(OkHttpClient httpClient, boolean includeRawResponses, String... urls)
```
Please let me know if there is anyway to overcome this in order not to change the position of the parameter.

### CODE REVIEW

File: core/src/main/java/org/web3j/protocol/http/HttpService.java [hunk 1]
```diff
@@ -154,6 +169,7 @@ protected InputStream performIO(String request) throws IOException {
         }
     }
 
+
```
Comment (iikirilov) [inline core/src/main/java/org/web3j/protocol/http/HttpService.java]: unnecessary whitespace

File: core/src/main/java/org/web3j/protocol/http/HttpService.java [hunk 2]
```diff
@@ -70,41 +70,43 @@
...
+        this(httpClient, includeRawResponses, DEFAULT_URL);
     }
 
-    private HttpService(String url, OkHttpClient httpClient) {
-        this(url, httpClient, false);
+    private HttpService(OkHttpClient httpClient, String url) {
+        this(httpClient, false, url);
     }
 
     public HttpService(String url) {
```
Comment (iikirilov) [inline core/src/main/java/org/web3j/protocol/http/HttpService.java]: You will need a constructor `HttpService(String... urls)` too

File: core/src/main/java/org/web3j/protocol/http/HttpService.java [hunk 3]
```diff
@@ -70,41 +70,43 @@
...
     }
 
     public HttpService(OkHttpClient httpClient, boolean includeRawResponses) {
-        this(DEFAULT_URL, httpClient, includeRawResponses);
+        this(httpClient, includeRawResponses, DEFAULT_URL);
     }
 
-    private HttpService(String url, OkHttpClient httpClient) {
-        this(url, httpClient, false);
+    private HttpService(OkHttpClient httpClient, String url) {
```
Comment (iikirilov) [inline core/src/main/java/org/web3j/protocol/http/HttpService.java]: You will need `HttpService(OkHttpClient httpClient, String... urls)`

File: core/src/main/java/org/web3j/protocol/http/HttpService.java [hunk 4]
```diff
@@ -70,41 +70,43 @@
...
     }
 
     public HttpService(String url) {
-        this(url, createOkHttpClient());
+        this(createOkHttpClient(), url);
     }
 
-    public HttpService(String url, boolean includeRawResponse) {
-        this(url, createOkHttpClient(), includeRawResponse);
+    public HttpService(boolean includeRawResponse, String url) {
```
Comment (iikirilov) [inline core/src/main/java/org/web3j/protocol/http/HttpService.java]: You will need `HttpService(boolean includeRawResponse, String... urls)`

File: core/src/main/java/org/web3j/protocol/http/HttpService.java [hunk 5]
```diff
@@ -70,41 +70,43 @@
...
+
+    private int counter = 0;
 
-    private final boolean includeRawResponse;
+    private boolean includeRawResponse;
 
     private HashMap<String, String> headers = new HashMap<>();
 
-    public HttpService(String url, OkHttpClient httpClient, boolean includeRawResponses) {
+    public HttpService(OkHttpClient httpClient, boolean includeRawResponses, String... urls) {
```
Comment (iikirilov) [inline core/src/main/java/org/web3j/protocol/http/HttpService.java]: You will need `HttpService(OkHttpClient httpClient, boolean includeRawResponses, String url)` too

File: core/src/main/java/org/web3j/protocol/http/HttpService.java [hunk 6]
```diff
@@ -199,6 +215,14 @@ public void addHeaders(Map<String, String> headersToAdd) {
         return headers;
     }
 
+    public String next() {
```
Comment (iikirilov) [inline core/src/main/java/org/web3j/protocol/http/HttpService.java]: call it `nextUrl()` - I think it should be a private method also.

File: core/src/main/java/org/web3j/protocol/http/HttpService.java [hunk 7]
```diff
@@ -131,13 +133,26 @@ protected InputStream performIO(String request) throws IOException {
...
+        try {
+            httpRequest = new okhttp3.Request.Builder()
+                    .url(url)
+                    .headers(headers)
+                    .post(requestBody)
+                    .build();
+        } catch (IllegalStateException e) {
+            url = next();
+            if (url != null) {
+                httpRequest = new okhttp3.Request.Builder()
```
Comment (iikirilov) [inline core/src/main/java/org/web3j/protocol/http/HttpService.java]: encapsulate in `private okhttp3.Request buildHttpRequest(...)` instead of copy pasta

File: core/src/main/java/org/web3j/protocol/http/HttpService.java [hunk 8]
```diff
@@ -131,13 +133,26 @@ protected InputStream performIO(String request) throws IOException {
...
+                        .url(url)
+                        .headers(headers)
+                        .post(requestBody)
+                        .build();
+            } else {
```
Comment (iikirilov) [inline core/src/main/java/org/web3j/protocol/http/HttpService.java]: it is good to do `log.error("message", e)` before rethrowing.

**Comment by iikirilov** [comment]
Also, tests required.

File: core/src/main/java/org/web3j/protocol/http/HttpService.java (hunk 7 above)
Comment (sebaraba) [reply core/src/main/java/org/web3j/protocol/http/HttpService.java]: You mean moving the check implementation inside the `okhttp3.Rquest` class?

Comment (iikirilov) [reply core/src/main/java/org/web3j/protocol/http/HttpService.java]: No. I mean write this method in HttpService class so you don't duplicate code.

**Comment by iikirilov** [comment]
You have checkstyle violations

**Comment by iikirilov** [comment]
> * [ ]  Also `includeRawResponses` and `url` were final before my changes. Is there a way to keep it final after the changes ?

You should be able to keep them as final.

**Comment by fcorneli** [comment]
This feature will also mess with the nonce cache... and hence fast transactions.

**Comment by iikirilov** [comment]
This feature has been dropped
